File: src/dotnet/Src/ButterflyNet.Parser/Strategies/OpenParagraphStrategy.cs
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	[Exclude]
	public class OpenParagraphStrategy : ScopeDrivenStrategy {
		public OpenParagraphStrategy() {
			AddSatisfier<CanNestParagraphSatisfier>();
		}

		protected override void DoExecute(ParseContext context) {
			OpenScope(new ParagraphScope(), context);
		}

		private class CanNestParagraphSatisfier : ISatisfier {
			public bool IsSatisfiedBy(ParseContext context) {
				return context.Scopes.IsEmpty() || context.Scopes.Peek().CanNestParagraph;
			}
		}
	}
}

File: src/js/src/strategies/OpenParagraphStrategy.js
function OpenParagraphStrategy() {
	OpenParagraphStrategy.$parent.call(this);

	this.addSatisfier(function(context) {
		return context.scopes.isEmpty() || context.scopes.peek().canNestParagraph;
	});

	this.doExecute = function(context) {
		this.openScope(new ParagraphScope(), context);
	};
}

extend(ScopeDrivenStrategy, OpenParagraphStrategy);